src/prefect/server/services/base.py

```diff
@@ -89,6 +94,32 @@ def enabled_services(cls) -> list[type[Self]]:
         """Get list of enabled service classes"""
         return [svc for svc in cls.all_services() if svc.enabled()]
 
+    @classmethod
+    @asynccontextmanager
+    async def running(cls) -> AsyncGenerator[None, None]:
+        """A context manager that runs enabled services on entry and stops them on
+        exit."""
+        service_tasks: dict[Service, asyncio.Task[None]] = {}
+        for service_class in cls.enabled_services():
+            service = service_class()
+            service_tasks[service] = asyncio.create_task(service.start())
+
+        try:
+            yield
+        finally:
+            await asyncio.gather(*[service.stop() for service in service_tasks])
+            await asyncio.gather(*service_tasks.values(), return_exceptions=True)
+
+    @classmethod
+    async def run_services(cls) -> NoReturn:
+        """Run enabled services until cancelled."""
+        async with cls.running():
+            try:
+                while True:
+                    await asyncio.sleep(1)
```

Comment:
I couldn't think of another way to just sleep forever, am I forgetting something obvious?

Reply:
This looks like a good place to use a `TaskGroup` from `anyio`. You could also gather all the service tasks and await them.

Reply:
I wanted to keep the tasks tucked away in `running` but I could yield them back out, will do that...

Reply:
Hmm, any approach that has me awaiting for the tasks here (including using a task group) kinda locks up trying to stop services when exiting the context manager.  I think for this "run indefinitely until you're cancelled" thing, the sleep loop might be the most straightforward

Reply:
I think if you implement `run_services` like this:
```python
    @classmethod
    async def run_services(cls) -> NoReturn:
        """Run enabled services until cancelled."""
        async with cls.running():
            forever = asyncio.get_running_loop().create_future()
            try:
                await forever
            except asyncio.CancelledError:
                logger.info("Received cancellation, stopping services...")
```
You'll get what you're looking for. That future will never resolve and keep the context open.

Reply:
Oh that's perfect!  will use that

Reply:
Yeah that worked great!